File: frontend/web/js/facts.js
var Facts = function() {

    return {
        //Determine age
        age: function(bdate) {
            var today = new Date()
            return today.getFullYear() - bdate.getFullYear()
        },

        //Body mass index
        //kg and m
        bmi: function(user) {
            var height = user.height / 100
            return user.weight / (height * height)
        },

        //BMI compared to country averae
        primeBmi: function(bmi, average) {
            return bmi / average
        },

        //Body fat percentage, U.S. Navy method
        //cm
        bfp: function(user) {
            if (user.gender == 'm')
                return 86.010 * Math.log(user.abdomen - user.neck) - 70.041 * Math.log(user.height) + 30.30
            else
                return 163.205 * Math.log(user.abdomen + user.hip - user.neck) - 97.684 * Math.log(user.height) - 104.912
        },

        //Basal Metabolic Rate, Harris-Benedict revised
        //kg and cm
        bmr: function(user) {
            var uage = 20//age(user.birthdate)
            if (user.gender)
                return 88.362 + 13.397 * user.weight + 4.799 * user.height - 5.677 * uage
            else
                return 447.593 + 9.247 * user.weight + 3.098 * user.height - 4.330 * uage
        },

        lifestyleCoef: function(user) {
            switch (user.lifestyle)
            {
            case 0: return 1.2
            case 1: return 1.375
            case 2: return 1.4625
            case 3: return 1.550
            case 4: return 1.6375
            case 5: return 1.725
            case 6: return 1.9
            }
        },
    }
}();

File: backend-ec2/facts.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

import math
import datetime

# BMR adjustment
lifestyle_coef = {
    0: 1.2,
    1: 1.375,
    2: 1.4625,
    3: 1.550,
    4: 1.6375,
    5: 1.725,
    6: 1.9,
}

# Body mass index
# kg and m
def bmi(height, weight):
    height = height / 100
    return weight / (height * height)

# BMI compared to country average
def prime_bmi(bmi, average):
    return bmi / average

# Body fat percentage, U.S. Navy method
# cm
def bfp(gender, height, abdomen, neck, hip=0):
    if gender:
        return 86.010 * math.log(abdomen - neck, 10) - 70.041 * math.log(height, 10) + 36.76
    else:
        return 163.205 * math.log(abdomen + hip - neck, 10) - 97.684 * math.log(height, 10) - 78.387

# Basal Metabolic Rate, Harris-Benedict revised
# kg and cm
def bmr(gender, age, height, weight):
    if gender:
        return 88.362 + 13.397 * weight + 4.799 * height - 5.677 * age
    else:
        return 447.593 + 9.247 * weight + 3.098 * height - 4.330 * age

def calculate_index(user, body):
    user_age = datetime.datetime.utcnow() - user['birthdate']
    user_age = user_age.days / 365

    bmi_value = bmi(body['height'], body['weight'])
    bfp_value = 0.0
    if body.get('abdomen', None):
        bfp_value = bfp(user['gender'], body['height'], body['abdomen'], body['neck'], body['hip'])
        bfp_value = bfp_value if bfp_value > 0 else 0
    bmr_value = bmr(user['gender'], user_age, body['height'], body['weight'])
    
    base_value = body['weight'] * (1 - bfp_value / 100.0)

    energy = bmr_value * lifestyle_coef[user['lifestyle']]
    protein = base_value * 5 # Enought potein to gain muscles
    carbohydrate = base_value # Low-carbs diet
    lipid = (energy - protein * 4 - carbohydrate * 9) / 4 # Rest calories

    index = {
        'bmi': bmi_value,
        'bfp': bfp_value,
        'bmr': bmr_value,
    }

    nutrition = {
        'energy': energy,
        'protein': protein,
        'lipid': lipid,
        'carbohydrate': carbohydrate
    }

    return (index, nutrition)